### DISCUSSION TITLE
NuiLine render singature potentialy incorrect

### DISCUSSION THREAD

**Discussion by vuki656** [MunifTanjim/nui.nvim#87] [Q&A] [answered]
Hey, is the [signature](https://github.com/MunifTanjim/nui.nvim/tree/main/lua/nui/line#linerender) for the render function incorrect? It has a namespace param that I assume shouldn't be there.

Also, it mentions `linenr_start`, does that mean on which line to start the render or on which column on the line to start render?

**Comment by MunifTanjim** [comment, marked as answer]
> Hey, is the signature for the render function incorrect? It has a namespace param that I assume shouldn't be there.

Nope, it's correct. `ns_id` is there intentionally. Namespace ID is used for highlighting. If you don't want to provide a custom namespace id, you can pass `-1` and it'll use the default namespace id (https://github.com/MunifTanjim/nui.nvim/blob/ec1028b92e26c2eaa36a09ccca492290b10da964/lua/nui/utils/init.lua#L71).

> Also, it mentions `linenr_start`, does that mean on which line to start the render or on which column on the line to start render?

`NuiLine` represent a whole line (start to end) on a buffer. So `linenr_start` is line to start the render.

If you only pass `linenr_start`, NuiLine will replace that line.
If you pass both `linenr_start, linenr_end` and they are different, NuiLine will replace the lines within that range.
If you pass both `linenr_start, linenr_end` and they are same, NuiLine will insert a new line at that line number.

**Reply by vuki656** [reply]
In the snippet example in the doc, there is this
```lua
local bufnr, linenr = 0, 1
line:render(bufnr, linenr)
```

If the second param to `render` is namespace id, is that example wrong?

I'm a little confused here :smile: 

---

What I'm trying to do is just render some text on the first line in the buffer and am using this assuming the first param is buffer number, the second one is namespace id and the third one is the line where to put it on.

```lua
    local title = Line({ Text("Press ? for help") })
    title:render(popup.bufnr, -1, 0)
```

But then I get the following error

![image](https://user-images.githubusercontent.com/8104698/147851506-1ab18e29-92b2-43af-aea4-809fcd22b727.png)

---

Also, unrelated to this, how can I center the text in the line?

**Reply by MunifTanjim** [reply]
> If the second param to render is namespace id, is that example wrong?

Ah, yes. That example ~is~ was wrong and ~needs to be updated~ fixed now. 😅 

> But then I get the following error

`linenr` is 1-indexed. You'll need to do `title:render(popup.bufnr, -1, 1)` for the first line.

---

> Also, unrelated to this, how can I center the text in the line?

You'll need to know about the total length and add spaces on the left and right. For example:

```lua
function calculate_gap_width(align, total_width, text_width())
  local gap_width = total_width - text_width
  if align == "left" then
    return 0, gap_width
  elseif align == "center" then
    return math.floor(gap_width / 2), math.ceil(gap_width / 2)
  elseif align == "right" then
    return gap_width, 0
  end

  error("invalid value align=" .. align)
end

local total_w = 30
local title = Text("Press ? for help")
local left_gap_w, right_gap_w = calculate_gap_width("center", total_w, title:width)

local line = Line()
if left_gap_w > 0 then
  line:append(string.rep(" ", left_gap_w))
end
line:append(title)
if right_gap_w > 0 then
  line:append(string.rep(" ", right_gap_w))
end
line:render(popup.bufnr, -1, 1)
```
